extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,73 @@
  * limitations under the License.
  */
 
-export class AmpViewerIntegration {
+import {Messaging} from './messaging.js';
+import {listen, unlisten} from '../../../src/event-helper';
+import {viewerPromiseForDoc} from '../../../src/viewer';
+import {user} from '../../../src/log';
+
+
+class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} win */
+    this.win_ = win;
+  }
+
+  init() {
+    let viewer;
+    this.getViewer()
+    .then(viewerParam => {
+      viewer = viewerParam;
+      return this.getHandshakePromise(viewer);
+    })
+    .then(viewerOrigin => {
+      const messaging = new Messaging(this.win_.parent, viewerOrigin,
+          function(type, payload, awaitResponse) {
+            return viewer.receiveMessage(type, payload, awaitResponse);
+          }, this.win_.location.href);
+      viewer.setMessageDeliverer(function(type, payload, awaitResponse) {
+        return messaging.sendRequest(type, payload, awaitResponse);
+      }, viewerOrigin);
+    });
+  }
+
+  /**
+   * @return {!Promise<!../../../src/service/viewer-impl.Viewer>}
+   */
+  getViewer() {
+    return viewerPromiseForDoc(this.win_.document);
+  }
+
+  /**
+   * @param {!../../../src/service/viewer-impl.Viewer} viewer
+   */
+  getHandshakePromise(viewer) {
+    const win = this.win_;
+
+    return new Promise(function(resolve) {
+      const unconfirmedViewerOrigin = viewer.getParam('viewerorigin');
+      user().assert(unconfirmedViewerOrigin,
+              'Expected viewer origin must be specified!');
+
+      const listener = function(event) {
+        if (event.origin == unconfirmedViewerOrigin &&
+                event.data == 'amp-handshake-response' &&
+                (!event.source || event.source == win.parent)) {
+          unlisten(win, 'message', listener);
+          resolve(event.origin);
+        }
+      };
+
+      listen(win, 'message', listener);
+
+      win.parent./*OK*/postMessage('amp-handshake-request',
+          unconfirmedViewerOrigin);
+    });
+  }
 }
+
+new AmpViewerIntegration(AMP.win).init();
```

Comment:
Let's wrap it into `AMP.extension()`. This way we don't need to worry about it self-installing for tests.